lib/src/merged_tree.rs

```diff
@@ -253,3 +274,148 @@ fn merge_tree_values(
         }
     }
 }
+
+pub struct ConflictEntriesNonRecursiveIterator<'a> {
+    merged_tree: &'a MergedTree,
+    candidate_names: Vec<&'a RepoPathComponent>,
+    next_pos: usize,
+}
+
+impl<'a> ConflictEntriesNonRecursiveIterator<'a> {
+    fn new(merged_tree: &'a MergedTree) -> Self {
+        // Use all basenames that changed between any pair of trees as candidates
+        let candidate_names = match merged_tree {
+            MergedTree::Legacy(tree) => tree
+                .entries_non_recursive()
+                .filter(|entry| matches!(entry.value(), &TreeValue::Conflict(_)))
+                .map(|entry| entry.name())
+                .collect(),
+            MergedTree::Merge(conflict) => {
+                if conflict.is_resolved() {
+                    vec![]
+                } else {
+                    conflict
+                        .removes()
+                        .iter()
+                        .chain(conflict.adds().iter())
+                        .flat_map(|tree| tree.data().names())
+                        .unique()
+                        .sorted_unstable()
+                        .collect()
+                }
+            }
+        };
+        ConflictEntriesNonRecursiveIterator {
+            merged_tree,
+            candidate_names,
+            next_pos: 0,
+        }
+    }
+}
+impl<'a> Iterator for ConflictEntriesNonRecursiveIterator<'a> {
+    type Item = (&'a RepoPathComponent, Conflict<Option<TreeValue>>);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while self.next_pos < self.candidate_names.len() {
+            let basename = self.candidate_names[self.next_pos];
+            self.next_pos += 1;
+            match self.merged_tree.value(basename) {
+                MergedTreeValue::Resolved(_) => {}
+                MergedTreeValue::Conflict(conflict) => {
+                    return Some((basename, conflict));
+                }
+            }
+        }
+        None
+    }
+}
+
+struct ConflictsDirItem {
+    entry_iterator: ConflictEntriesNonRecursiveIterator<'static>,
+    // On drop, tree must outlive entry_iterator
+    tree: Pin<Box<MergedTree>>,
+}
+
+impl ConflictsDirItem {
+    fn new(tree: MergedTree) -> Self {
+        let tree = Box::pin(tree);
+        let entry_iterator = ConflictEntriesNonRecursiveIterator::new(&tree);
```

Comment:
Interesting! I've fixed it here. Feel free to fix it in the other places. Otherwise I'll try to remember to do that when I get back from vacation in a few days.